frontend/src/components/data-table/column-pinning/feature.ts

```diff
@@ -0,0 +1,194 @@
+          };
+        }
+
+        // If prevValue isn't false and position is false,
+        // remove the column from the specified side
+        if (prevValue !== false && !position) {
+          newState = {
+            ...old,
+            [prevValue]: old[prevValue]?.filter((id) => id !== column.id),
+          };
+        }
+
+        // If prevValue isn't false and position is different,
+        // move the column to the new side
+        if (prevValue !== false && position && prevValue !== position) {
+          newState = {
+            ...old,
+            [prevValue]: old[prevValue]?.filter((id) => id !== column.id),
+            [position]: [...(old[position] || []), column.id],
+          };
+        }
+
+        // If only "select" column is pinned to the left, unpin it
+        if (
+          position !== "left" &&
+          newState.left?.length === 1 &&
+          newState.left[0] === "select"
+        ) {
+          // unpin the 'select' column
+          table.getColumn("select")?.pin(false);
+
+          // remove the 'select' column from the left side
+          newState = {
+            ...newState,
+            left: [],
+          };
+        }
+
+        return newState;
+      };
+      table.options.onColumnPinningChange?.(safeUpdater);
+    };
+
+    // toggle the column's pinning state
+    column.toggleColumnPinning = (position: ColumnPinningPosition) => {
+      // get the column pinning state
+      const { columnPinning } = table.getState();
+
+      // if the column is being pinned to the left,
+      // but the 'select' column is not pinned to the left,
+      // pin the 'select' column to the left
+      if (position === "left" && !columnPinning.left?.includes("select")) {
+        // pin the 'select' column to the left
+        const selectColumn = table.getColumn("select");
+        if (selectColumn) {
+          // always setColumnPinning first then pin
+          selectColumn.setColumnPinning("left");
+          selectColumn.pin("left");
+        }
+      }
+      column.setColumnPinning(position);
+      column.pin(position);
+    };
+
+    // get the column's common pinning styles
+    column.getCommonPinningStyles = (): CSSProperties => {
+      // get the column's pinning state
+      const isPinned = column.getIsPinned();
+
+      // get if the column is the last pinned column on the left
+      const isLastLeftPinnedColumn =
+        isPinned === "left" &&
+        column.getIsLastColumn("left") &&
+        // don't apply the box shadow to the 'select' column
+        column.id !== "select";
+
+      // get if the column is the first pinned column on the right
+      const isFirstRightPinnedColumn =
+        isPinned === "right" && column.getIsFirstColumn("right");
+
+      return {
+        // apply a box shadow to the column if
+        // it is the last pinned column on the left
+        // or the first pinned column on the right
+        boxShadow: isLastLeftPinnedColumn
```

Comment:
it could be nice if we can use classnames instead of styles - so we can leverage our tailwind css.

could we do `shadow-inner` or `shadow-sm` here?

maybe we could add `column.getCommonPinningClassNames` as well.

Reply:
or maybe this is better:
```ts
getCommonPinningClassNameAndStyles() => {
  return {
   className: "...",
   styles: {},
  }
}
```

Reply:
just tried this and it may not work. at least instead of `gray`, could we do `var(--base-shadow)`